### PR TITLE
Custom Functions Optional Parameters

### PR DESCRIPTION
**Pull request by mscharlock**
1. Changing title according to feedback from Reeza 

2. Documentation write up for CF optional parameters

### CODE REVIEW

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C26%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811261812433372-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=jLHr57Uu7P1zPFKauHqq8VdL3Xh9Ij1BGFqyrqDMB0A%3D&st=2018-11-26T18%3A08%3A56Z&se=2018-12-27T18%3A13%3A56Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C26%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811261813520773-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=AIX8mmrsc3XcrsbuwkMo4aihiOifvUvytEU4CnzqPw8%3D&st=2018-11-26T18%3A10%3A13Z&se=2018-12-27T18%3A15%3A13Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/custom-functions-json.md [hunk 1]
```diff
@@ -133,6 +133,9 @@ The `parameters` property is an array of parameter objects. The following table
 |  `dimensionality`  |  string  |  No  |  Must be either **scalar** (a non-array value) or **matrix** (a 2-dimensional array).  |
 |  `name`  |  string  |  Yes  |  The name of the parameter. This name is displayed in Excel's intelliSense.  |
 |  `type`  |  string  |  No  |  The data type of the parameter. Can be **boolean**, **number**, **string**, or **any**, which allows you to use of any of the previous three types. If this property is not specified, the data type defaults to **any**. |
+|  `optional`  | string | No | If **true**, parameter will be labelled optional. |
+
+[!NOTE] Any as the default value for `type` and optional parameters are currently available, but you may experience linting errors and not see optional parameters in Excel's formula UI when calling a function. This is projected to change in December of 2018.
```
Comment (kbrandl) [inline docs/excel/custom-functions-json.md]: It's not clear to me what this note is trying to say, as currently written.  Is it intended to convey something like this? -> 

If the `type` property of an optional parameter is either not specified or is specified and set to **any**, you may notice issues such as linting errors in your IDE and optional parameters not being displayed when the function is being entered into a cell in Excel. This is projected to change in December 2018.

Comment (mscharlock) [reply docs/excel/custom-functions-json.md]: Yes, trying to sort of say what you have. This is kind of a weird situation where the schema (which I think provides the intelliSense) is not currently up to date but the feature is already supported. So it appears as an error when you try to use it, even though it works.

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C29%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811291804518058-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=QUe4pMp5R0%2BL1iSOD4RytMHC6GBCh9xco9zCXjaoyRQ%3D&st=2018-11-29T18%3A01%3A11Z&se=2018-12-30T18%3A06%3A11Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C29%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811291810433867-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=ziAwIGb0%2FZXBuGbSyL7ZZWcLpRTjYkorisvUW1B4PIU%3D&st=2018-11-29T18%3A06%3A55Z&se=2018-12-30T18%3A11%3A55Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by mscharlock** [comment]
@kbrandl do you mind looking these over when you have a minute? Thanks!

File: docs/excel/custom-functions-best-practices.md [hunk 2]
```diff
@@ -1,12 +1,12 @@
 ---
-ms.date: 10/24/2018
+ms.date: 11/29/2018
 description: Learn best practices and recommended patterns for Excel custom functions.
 title: Custom functions best practices
 ---
 
-# Custom functions best practices (preview)
+# Custom functions design considerations (preview)
```
[the remark this answers is not in the document]
Comment (mscharlock) [reply docs/excel/custom-functions-best-practices.md]: Right! D'oh. Thanks

File: docs/excel/custom-functions-best-practices.md [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
-This article describes best practices for developing custom functions in Excel.
+This article describes design patterns and practices for developing custom functions in Excel.
```
[the remark this answers is not in the document]
Comment (mscharlock) [reply docs/excel/custom-functions-best-practices.md]: Ok good to know. Yes, this is a weird catch-all that we want to break out. My plan is to eventually separate out the debugging/testing stuff to it's own article but leave the mapping, optional params in this best practices one? Thoughts?

[the remark this answers is not in the document]
Comment (mscharlock) [reply docs/excel/custom-functions-best-practices.md]: Also it seems like there have been two schools of thought on this kind of thing: either break more complex topics into individual articles as you suggest or keep everything in one long overview topic with the assumption that people use the right hand nav to skip through. I'm kinda neutral and can see advantages to either side (and we don't have data on either approach). My thing about breaking them up is that it then makes me wonder what else ought to come out of the overview into their own pages (streaming & cancellable functions would be good candidates).

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811300038096662-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=Lri%2BknWz%2BqWTGoV7E5lsTbolwYp3aprd5gOQS7qj2wI%3D&st=2018-11-30T00%3A34%3A24Z&se=2018-12-31T00%3A39%3A24Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/custom-functions-best-practices.md [hunk 4]
```diff
@@ -125,6 +125,63 @@ Keep in mind the following best practices when creating custom functions in your
...
+function getWeatherReport(required, zipCode, dayOfWeek)
+{
+  if (zipCode === undefined) {
+      zipCode === 98052
+  }
+  if (dayOfWeek === undefined) {
+    dayOfWeek === "Wednesday"
+}
+    ... //After establishing parameters, this function would have additional code to get a weather report
+}
```
Comment (kbrandl) [inline docs/excel/custom-functions-best-practices.md]: Multiple issues with this code snippet:
- the first input parameter (`required`) is not used and doesn't make any sense in this context 
- incorrect use of triple-equals-signs for assigning value to a variable
- missing semi-colons
- inconsistent use of indentation: 4 spaces on some lines, 2 spaces on other lines
- closing brace on the second `if` statement is missing indentation
- awkward positioning and wording of code comment

Suggest replacing the code snippet with the following  (and then verifying what the snippet looks like on the staging site after you commit/push and the OPS build completes):

```
function getWeatherReport(zipCode, dayOfWeek)
{
  if (zipCode === undefined) {
    zipCode = "98052";
  }

  if (dayOfWeek === undefined) {
    dayOfWeek = "Wednesday";
  }

  // Get weather report for specified zipCode and dayOfWeek
  // ... 
}
```

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811300136258692-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=DIpS17j5f4ylFPEkYyqlVQ83Aeln2cUf4xZId5%2FItb4%3D&st=2018-11-30T01%3A32%3A37Z&se=2018-12-31T01%3A37%3A37Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/custom-functions-best-practices.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbrandl) [reply docs/excel/custom-functions-best-practices.md]: I use the [files](https://github.com/OfficeDev/office-js-docs-pr/pull/576/files) view to review changes, add comments, and review previous comments (resolved or not), etc.  I'm guessing that the "files" view drops comment threads that it considers 'Outdated', while [this view](https://github.com/OfficeDev/office-js-docs-pr/pull/576) shows all comments (including 'Outdated' ones).

File: docs/excel/custom-functions-best-practices.md (hunk 4 above)
Comment (kbrandl) [reply docs/excel/custom-functions-best-practices.md]: @mscharlock looks like you applied some of the suggested fixes to this [code sample](https://review.docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-best-practices?branch=michelle-optionalparams#declaring-optional-parameters), but it still needs a couple more tweaks. 

image

- add semicolon to the end of these lines:
    - `zipCode = "98052";`
    - `dayOfWeek = "Wednesday";`
- fix indentation (outdent by 2 spaces) for the last three lines before the final brace:
    - `}`
    - `// Get weather report for specified zipCode and dayOfWeek`
    - `// ...`

Comment (mscharlock) [reply docs/excel/custom-functions-best-practices.md]: Yes, sorry about this sample - I think I was rushing through things too quickly yesterday and had to get out the door to get home to meet someone so I didn't get all the way through (hence, this is still unresolved). Will fix this am.

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C11%5C30%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201811301801097658-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=Bb%2BJBuFkQyIqzRldYG%2BB%2B60P81fahecaaxZTZHDU%2B7s%3D&st=2018-11-30T17%3A57%3A27Z&se=2018-12-31T18%3A02%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by o365devx** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/custom-functions-best-practices.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-best-practices.md) | :white_check_mark:Succeeded | |
[docs/excel/custom-functions-json.md](https://github.com/OfficeDev/office-js-docs-pr/blob/michelle-optionalparams/docs/excel/custom-functions-json.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C12%5C1%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201812010014125545-576%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=N1iFnQMmroROp62DqhDaeW%2FfWHF4mkWuH8MHshgIkzc%3D&st=2018-12-01T00%3A10%3A22Z&se=2019-01-01T00%3A15%3A22Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by kbrandl** [comment]
@mscharlock FYI, I just merged a PR into this branch to fix the remaining formatting issues with the code sample.

**Comment by mscharlock** [comment]
@kbrandl Super - Alex just explained to me why zips are strings. :)

**Comment by mscharlock** [comment]
@kbrandl Yes, that sounds like a good plan. At some point, we might be able to abandon this best practices one too, seeing as we will be moving debugging to a new topic. I'll plan some time next week to chat!

**Pull request merged**